File: hmis-engine-web/src/app/features/iam/roles/role.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable, inject } from '@angular/core';
import { Observable } from 'rxjs';

import { environment } from '../../../../environments/environment';
import { CreateRoleRequest, Privilege, Role } from './role.types';

@Injectable({ providedIn: 'root' })
export class RoleService {
  private readonly http = inject(HttpClient);
  private readonly base = `${environment.apiUrl}/iam`;

  list(): Observable<Role[]> { return this.http.get<Role[]>(`${this.base}/roles`); }
  findByUid(uid: string): Observable<Role> { return this.http.get<Role>(`${this.base}/roles/uid/${uid}`); }
  create(req: CreateRoleRequest): Observable<Role> { return this.http.post<Role>(`${this.base}/roles`, req); }
  replacePrivileges(uid: string, privileges: string[]): Observable<Role> {
    return this.http.put<Role>(`${this.base}/roles/uid/${uid}/privileges`, privileges);
  }

  listPrivileges(): Observable<Privilege[]> { return this.http.get<Privilege[]>(`${this.base}/privileges`); }
}

File: hmis-engine-api/src/main/java/com/otapp/hmis/engine/iam/application/RoleService.java
package com.otapp.hmis.engine.iam.application;

import com.otapp.hmis.engine.common.error.ConflictException;
import com.otapp.hmis.engine.common.error.NotFoundException;
import com.otapp.hmis.engine.iam.application.dto.CreateRoleRequest;
import com.otapp.hmis.engine.iam.application.dto.PrivilegeDto;
import com.otapp.hmis.engine.iam.application.dto.RoleDto;
import com.otapp.hmis.engine.iam.domain.Privilege;
import com.otapp.hmis.engine.iam.domain.PrivilegeRepository;
import com.otapp.hmis.engine.iam.domain.Role;
import com.otapp.hmis.engine.iam.domain.RoleRepository;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
public class RoleService {

    private final RoleRepository roleRepository;
    private final PrivilegeRepository privilegeRepository;

    @Transactional
    public RoleDto create(CreateRoleRequest request) {
        if (roleRepository.existsByName(request.name())) {
            throw new ConflictException("Role already exists: " + request.name());
        }
        Role role = new Role(request.name(), request.description());
        attachPrivileges(role, request.privileges());
        roleRepository.save(role);
        return IamMapper.toDto(role);
    }

    @Transactional(readOnly = true)
    public List<RoleDto> list() {
        return roleRepository.findAll().stream()
                .map(IamMapper::toDto)
                .toList();
    }

    @Transactional(readOnly = true)
    public RoleDto findByUid(String uid) {
        return roleRepository.findByUid(uid)
                .map(IamMapper::toDto)
                .orElseThrow(() -> new NotFoundException("Role not found: " + uid));
    }

    @Transactional
    public RoleDto replacePrivileges(String uid, Set<String> privilegeNames) {
        Role role = roleRepository.findByUid(uid)
                .orElseThrow(() -> new NotFoundException("Role not found: " + uid));
        role.getPrivileges().clear();
        attachPrivileges(role, privilegeNames);
        return IamMapper.toDto(role);
    }

    @Transactional(readOnly = true)
    public List<PrivilegeDto> listPrivileges() {
        return privilegeRepository.findAll().stream()
                .map(IamMapper::toDto)
                .toList();
    }

    private void attachPrivileges(Role role, Set<String> names) {
        if (names == null || names.isEmpty()) {
            return;
        }
        Set<Privilege> resolved = new HashSet<>();
        for (String name : names) {
            Privilege priv = privilegeRepository.findByName(name)
                    .orElseThrow(() -> new NotFoundException("Privilege not found: " + name));
            resolved.add(priv);
        }
        resolved.forEach(role::grant);
    }
}
